tests/integration_tests/functional/test_rng.py

```diff
@@ -122,3 +123,111 @@ def test_rng_snapshot(test_microvm_with_rng, microvm_factory):
     )
 
     check_entropy(new_vm.ssh)
+
+
+def _get_percentage_difference(measured, base):
+    """Return the percentage delta between the arguments."""
+    if measured == base:
+        return 0
+    try:
+        return (abs(measured - base) / base) * 100.0
+    except ZeroDivisionError:
+        # It means base and only base is 0.
+        return 100.0
+
+
+def _throughput_units_multiplier(units):
+    """
+    Parse the throughput units and return the multiplier that would
+    translate the corresponding value to Bytes/sec
+    """
+    if units == "kB/s":
+        return 1000
+
+    if units == "MB/s":
+        return 1000 * 1000
+
+    if units == "GB/s":
+        return 1000 * 1000 * 1000
+
+    raise Exception("Unknown units")
+
+
+def _process_dd_output(out):
+    """
+    Parse the output of `dd` and return the achieved throughput in
+    KB/sec
+    """
+    report = out.splitlines()[-1].split(" ")
+    (value, units) = (report[-2], report[-1])
+
+    return float(value) * _throughput_units_multiplier(units) / 1000
+
+
+def _get_throughput(ssh, random_bytes):
+    """
+    Request `random_bytes` from `/dev/hwrng` and return the achieved
+    throughput in KB/sec
+    """
+    cmd = "dd if=/dev/hwrng of=/dev/null bs={} count=1".format(random_bytes)
+    exit_code, _, stderr = ssh.execute_command(cmd)
+    assert exit_code == 0, stderr.read()
+
+    # dd gives its output on stderr
+    return _process_dd_output(stderr.read())
+
+
+def _check_entropy_rate_limited(ssh, random_bytes, expected_kbps):
+    """
+    Ask for `random_bytes` from `/dev/hwrng` in the guest and check
+    that achieved throughput is within a 10% of the expected throughput.
+
+    NOTE: 10% is an arbitrarily selected limit which should be safe enough,
+    so that we don't run into many intermittent CI failures.
+    """
+    measured_kbps = _get_throughput(ssh, random_bytes)
+    assert (
+        _get_percentage_difference(measured_kbps, expected_kbps) <= 10
+    ), "Expected {} KB/s, measured {} KB/s".format(expected_kbps, measured_kbps)
+
+
+def _rate_limiter_id(rate_limiter):
+    """
+    Helper function to return a name for the rate_limiter to be
+    used as an id for parametrized tests.
+    """
+    size = rate_limiter["bandwidth"]["size"]
+    refill_time = rate_limiter["bandwidth"]["refill_time"]
+
+    return "{} KB/sec".format(size / (refill_time / 1000.0) / 1000.0)
```

Comment:
nit: size / (refill_time / 1000) / 1000 = (size * 1000 / refill_time) / 1000 = size / refill_time (since size is in bytes and refill_time is in millseconds, both units are off by the same factor of 1000 and we get kb/s)

Reply:
You are right! Adjusting